google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Subscriber.java

```diff
@@ -284,6 +289,94 @@ protected void doStop() {
     }
   }
 
+  private void startStreamingConnections() throws IOException {
+    synchronized (streamingSubscriberConnections) {
+      Credentials credentials = credentialsProvider.getCredentials();
+      CallCredentials callCredentials =
+          credentials == null ? null : MoreCallCredentials.from(credentials);
+
+      for (Channel channel : channels) {
+        SubscriberStub stub = SubscriberGrpc.newStub(channel);
+        if (callCredentials != null) {
+          stub = stub.withCallCredentials(callCredentials);
+        }
+        streamingSubscriberConnections.add(
+            new StreamingSubscriberConnection(
+                cachedSubscriptionNameString,
+                receiver,
+                ackExpirationPadding,
+                maxAckExtensionPeriod,
+                streamAckDeadlineSeconds,
+                ackLatencyDistribution,
+                stub,
+                flowController,
+                executor,
+                alarmsExecutor,
+                clock));
+      }
+      startConnections(
+          streamingSubscriberConnections,
+          new Listener() {
+            @Override
+            public void failed(State from, Throwable failure) {
+              // If a connection failed is because of a fatal error, we should fail the
+              // whole subscriber.
+              stopAllStreamingConnections();
+              try {
+                notifyFailed(failure);
+              } catch (IllegalStateException e) {
+                if (isRunning()) {
+                  throw e;
+                }
+                // It could happen that we are shutting down while some channels fail.
+              }
+            }
+          });
+    }
+
+    ackDeadlineUpdater =
+        executor.scheduleAtFixedRate(
+            new Runnable() {
```

Comment:
This is a somewhat large anonymous class - could you extract it out of this method?